src/homework04.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+# coding: utf8
+
+"""
+EPAM python q4 homework 04
+Sergey Streltsov 2019-11-12
+"""
+
+
+import math
+import time
+
+
+def save_history(f_name):
+    """History decorator
+
+    Create a decorator that stores the results of function calls and the
+    time it takes to get the result during program startup
+    """
+    history = list()
+
+    def wrapper(*args, **kwargs):
+        start_time = int(round(time.time() * 1000))
+        f_result = f_name(*args, **kwargs)
+        stop_time = int(round(time.time() * 1000))
+        est_time = stop_time - start_time
+        history.append({'name': f_name.__name__,
+                        'result': f_result,
+                        'time_ms': est_time})
+        print(history)
+        return f_result
+    return wrapper
+
+
+@save_history
+def count_factorial(x):
+    return math.factorial(x)
+
+
+def get_ranges(in_list):
+    """Collapsed list
+
+    Implement the get_ranges function which receives a non-empty list
+    of non-repeating integers, sorted in ascending order, which this
+    list “collapses”
+    """
+    last = None
+    raw_str = ''
+    for number in in_list:
+        if not (last is None):
+            if number - last == 1:
+                raw_str += ('-' + str(number))
+                last = number
+            else:
+                raw_str += ',' + str(number)
+                last = number
+        else:
+            raw_str += str(number)
+            last = number
+    new_list = list()
+    for block in raw_str.split(','):
+        if block.count('-') > 1:
+            new_block = '-'.join([block.split('-')[0], block.split('-')[-1]])
```

Comment:
Overall it looks like here we make redundant structure like num-num+1-num+2, other_num and than parse it with sometimes O(n**2) complexity. 
It looks it's exceptionally redundant as we have a lot of implicit parts (why do we need all these -, why do we need 0 and -1 elements. ) 
Overall, it looks too complex for this task.

Reply:
the structure was: num1-num2-num3,num4,num5,num6-num7.... ect. And after I parse the string and eliminate body from blocks between commas. Now I implemented v2 of algo, but I mean this v2 not so pythonic ) because using list indexes...